Convert JoystickButtons to int & increase available **REQUIRES TESTING**

leezer3 opened the pull request:
This is a continuation of this pull:
https://github.com/opentk/opentk/pull/344

**Changes:**
Joystick buttons have been changed to ints.
Buttons up-to and including 32 may be accessed via the old Joystick.Buttons enum, greater than this will print a debug message.
There is a hard limit of 64 buttons- Bitshifting problem....

**Notes:**
I have only tested Windows at the minute. Will do Linux, SDL2 and OS-X tonight, but it should work....
I also haven't cleaned up the code as well as I'd like just yet.

leezer3 commented:
Further note:
I had to convert the buttons storage variable to an In64, and fiddle with the bitshifting a lot. This definitely works under Windows, but not tested at all under other platforms.....

cra0zy commented:
Please go over your changes and make sure you are using spaces and not tabs.

leezer3 commented:
Commit has been amended to fix spacing issues.

Have also tested Linux and OS-X, which work, and the code looks good, but I can't test with anything greater than 32 buttons, as vJoy (Virtual joystick I've been testing with) doesn't work on anything other than Windows, and VMWare doesn't want to connect it either.

leezer3 commented:
I think this diff link works:
https://github.com/opentk/opentk/pull/380/files#diff-032dbd2ebf4169534ab5bac656ac637eR44

This is where I've set the max buttons. The HID driver needs to know the maxiumum number of buttons we support, in order to set the page size.
At present, it still throws an argument out of range exception if this limit is exceeded- I don't think there's much that can be done with this without bashing around the HID driver, which is well out of my comfort zone.
As noted above, it ought to be possible to provide a setter for this via ToolkitOptions so that if people really want zillions of buttons we can oblige.

cra0zy commented on Source/OpenTK/Input/JoystickState.cs:
```diff
@@ -238,22 +287,22 @@ internal void SetAxis(JoystickAxis axis, short value)
         internal void ClearButtons()
         {
-            buttons = 0;
+            buttons = (Int64)0;
```
Pointless cast.

cra0zy commented on Source/OpenTK/Input/JoystickState.cs:
```diff
@@ -108,18 +125,50 @@ public JoystickHatState GetHat(JoystickHat hat)
+        /// <param name="button">The <see cref="JoystickButton"/> to query.</param>
+        public bool IsButtonUp(int button)
+        {
             return (buttons & (1 << (int)button)) == 0;
```
Pointless cast.

cra0zy commented on Source/OpenTK/Input/JoystickState.cs:
```diff
@@ -41,13 +41,13 @@ public struct JoystickState : IEquatable<JoystickState>
         // then we'll need to increase these limits.
         internal const int MaxAxes = (int)JoystickAxis.Last + 1;
-        internal const int MaxButtons = (int)JoystickButton.Last + 1;
+        internal const int MaxButtons = 128;
```
With the code you are doing bellow, the MaxButton count should be 64, ie. `(Int64)1 << index` will get out of bounds and loop back around after 63

cra0zy commented on Source/OpenTK/Input/JoystickState.cs:
```diff
@@ -41,13 +41,13 @@ public struct JoystickState : IEquatable<JoystickState>
 
         int packet_number;
-        int buttons;
+        Int64 buttons;
```
Please use `long` instead of `Int64`

leezer3 commented:
Have amended.

Some notes:
The cast to **Int64** (Your first line note) seems to be required to get bitshifting to work right. If I don't cast to Int64, then things go pear shaped....
I suspect that the runtime is storing the zero as a standard int unless forced to do otherwise :)

Second one would appear to be broken anyway, as I forgot the casts to Int64 to get bitshifting to work right, have fixed this properly......

Third point-
Thinking about it, that seems logical, but it appears to work for me with 128 buttons defined.....
I've reverted to 64 buttons, will need to do some investigating as to why 128 works OK at this end.

cra0zy commented on Source/OpenTK/Input/JoystickState.cs:
```diff
@@ -108,22 +125,54 @@ public JoystickHatState GetHat(JoystickHat hat)
+        /// <param name="button">The <see cref="JoystickButton"/> to query.</param>
+        public bool IsButtonUp(int button)
+        {
+			return (buttons & (long)((long)1 << button)) == 0;
```
Tabs >:|

leezer3 commented:
Fudge, will fix that in a minute, must test as well....

cra0zy commented:
> will need to do some investigating as to why 128 works OK at this end.

2 buttons will have the exact same value, think about it, you cannot put in 128 bits inside 64 bits.

leezer3 commented:
Yeah, I know. That seems almost too logical!

I cooked up a rather simple test app that just printed the button pressed to console, and that was working OK though.

It was almost certainly something I've done, but I haven't really played with bitshifting enough to be confident I haven't done something entirely stupid, hence why I add the very large testing required!

Hence why I've got no intention of being a maintainer :+1:

leezer3 commented:
Two more pointless bits of code, sorry :)

wcdeich4 commented:
[FreeGLUT ](http://freeglut.sourceforge.net/)  has pretty good support for joysticks on multiple platforms. That's a different language (pure C), but if anyone wants to compare the code, it might provide some useful ideas - just a suggestion if anyone is interested. 

@leezer3 thank you for all your hard work on this!!!!  If I ever get my flight simulator working, I definitely want to use the new OpenTK joystick functions :)

leezer3 commented:
I've got users bugging me :p
Three at the moment I know about, and i'm sure there are more.

Simmers tend to have large numbers of buttons on their sticks!

Back on topic a little- It wouldn't be much work to add a second long for buttons 64 to 128- Dunno how many the biggest 'real' (Common?) stick has.
It's a pity OpenTK targets .Net 2.0 as the minimum, as v 4.0 brings 128 bit integers. I suppose an alternative is to write our own type, but that's way OTT for joystick buttons.

wcdeich4 commented:
MonoDevelop now supports .NET 4.0, and I think there's even an update for supporting 4.5.  Is there any reason to be hindered by supporting 2.0??

leezer3 commented:
No comment.

From a relatively small-time developer's point of view though:
I don't necessarily think that wanting the **BigInteger** type for a single field is a particularly good reason for bumping the required version though.
You might also want to check what the .Net version support is like on some of your more esotoric platforms (ARM for example) before doing a major .Net version reqiurement bump.

cra0zy commented:
> MonoDevelop now supports .NET 4.0, and I think there's even an update for supporting 4.5. Is there any reason to be hindered by supporting 2.0??

Both mono and monodevelop supported .NET 4.0 for some time now... you've been living under a rock.

> From a relatively small-time developer's point of view though:
> I don't necessarily think that wanting the BigInteger type for a single field is a particularly good reason for bumping the required version.
> You might also want to check what the .Net version support is like on some of your more esotoric platforms (ARM for example) before doing a major .Net version reqiurement bump.

Instead of using something like bind integers, we could just setup an array of int and use that instead.

The code would then be something like (for array length 4 and using 128 buttons and 32 bit int): `buttons[index / 4] < index % 4;`

wcdeich4 commented:
Forgive me if I've been under my rock too long, but why an array of length 4?

leezer3 commented:
I presume that was just an example....

I can't say the idea of variable length arrays of bitshifting especially thrills me, but I can't think of a better solution off the top of my head, other than something like an even larger array, which is yuck.

wcdeich4 commented:
The reason I said "Forgive me if I've been under my rock too long" is b/c it seems to me if you expect 512 to be the max number of buttons - why not just have an array of length 512?..... or am I missing something obvious?

cra0zy commented:
> The reason I said "Forgive me if I've been under my rock too long" is b/c it seems to me if you expect 512 to be the max number of buttons - why not just have an array of length 512?..... or am I missing something obvious?

That's too memory inefficient, in my example every bit is being used.

For example let's say you want to get 128 buttons states with an array. You can use bool so we have:
```c#
buttons = new bool[128];
```
bool takes 1 byte, so we have 128 * 1 = 128 bytes of memory used. On the other hand doing it with an int like I did above would have buttons looking like:
```c#
buttons = new int[4];
```
each int takes 4 bytes of memory so we have 4 * 4 = 16 bytes of memory.

leezer3 commented:
Not just memory-
Bitshifting means we can use a single integer comparison to check if any of 64 buttons are down `buttons != 0`
Changing to an array of bools would require cycling through the whole array.

Back onto topic:
I think using **longs** as opposed to **ints** in the proposed array would probably be better, simply as that reduces the length of the array and the number of potential calculations.
Not 100% happy with it, as sticks with over 32 buttons are somewhat of an edge case, and instancing an array of which only one member gets get used most of the time feels a little iffy.

```
public bool IsButtonDown(int button)
        {
            return (buttons[button / 4] & (long)((long)1 << buttons[button / 4])) != 0;
        }
```
I think that should work. Looks like we'd only need to alter this in the joystick state, as everywhere else just calls on the button index.

Do you want me to update the pull in this style?

wcdeich4 commented:
@cra0zy isn't this going to run in the computer's main memory? Modern RAM chips are measured in Gigabytes.  @leezer3 I know bit shifting is fast, but if we had If (boolArray[I] == true) the compiler usually converts that to an offset from the base address of the array - is that really slower?

cra0zy commented:
@wcdeich4 No offense, but you really are a noob when it comes to programming... google your questions instead of asking @leezer3 and me thousand of pointless question that any first year programming student knows!!!

leezer3 commented:
So, an array would appear to work as expected, but with one small hitch I don't exactly like.
The **IsAnyButtonDown** methodswould need to check all members of the array. 
Best I can come up with off the top of my head is a simple for-loop:
```C#
        public bool IsAnyButtonDown
        {
            get
            {
                // If any bit is set then a button is down.
	            for (int i = 0; i > buttons.Length; i++)
	            {
		            if (buttons[i] != 0)
		            {
			            return true;
		            }
	            }
	            return false;
            }
        }
```

Any better thoughts?
That's assuming a variable length array. 
If using an array of a fixed known length, then it would be possible to use an oversized version of the current return statement, but that doesn't really feel very nice either.....

leezer3 commented:
OK, so the above amended commit is a briefly tested WIP:

Other than the array, which I can't see an easy way around, it has the following issue:
I couldn't think of an easy (obvious) place to initialise the array. JoystickState hasn't got a constructor, so that's out. It'd probably be easiest to stick it into the initial joystick detection, but thoughts?

It's here as a temporary measure, but obviously doesn't want to stay there:
https://github.com/opentk/opentk/pull/380/commits/33a9ecf380b1565ed7a80252996af166f287a562#diff-032dbd2ebf4169534ab5bac656ac637eR305
The array length initialisation would probably be better changed to this:
```C#
buttons = new long[(MaxButtons / 64) +1];
```
Doing that though would require adding a place to set the max buttons dynamically, possibly via the configuration namespace as noted above.

wcdeich4 commented:
@leezer3 how's it coming?

leezer3 commented:
Hadn't forgotten it.....

Not 100% happy with the above commit, but would be interested in thoughts from anyone else.
What I've done:
Added a MaxJoystickButtons value to the ToolkitOptions and Configuration files.
When the toolkit is initiated, the maximum number of buttons is set to 32, unless set by the user (Tempted to change this to 128, but 32 is what's the current default.....)
Added a new constructor to JoystickState, and set all instances of it to use this constructor, which inits the array etc

The constructor is the part I'm not happy with, but I can't find a way round it.....

wcdeich4 commented:
@leezer3 thanks for your hard work.  Please keep it up; I want to use the joystick functionality in  a flight simulator someday :o)   BTW, if you wanted to use an oversized array, you could use if (array[0] || array[1] || ... array[511]) { return true; } else { return false; } --- but please do what you think is best

varon commented:
@leezer3 This seems to be conflicted with the current source changes. Reckon you could check that?

leezer3 commented:
Will do.

Having said that, it's not really something I'm 100% happy with, as the constructor was rather a hack :)

Will do a little more playing at the same time & see if I can get rid of it....

wcdeich4 commented:
@leezer3 - I just want to make sure you understand what @varon is talking about. There have been massive changes to the code in release 2.0 (large scale clean up & the examples are not even in the same solution) - the reason your pull request is no longer compatible is b/c it was based on 1.1
